sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala

```diff
@@ -636,3 +650,222 @@ class LoopStatementExec(
     body.reset()
   }
 }
+
+/**
+ * Executable node for ForStatement.
+ * @param query Executable node for the query.
+ * @param variableName Name of variable used for accessing current row during iteration.
+ * @param body Executable node for the body.
+ * @param label Label set to ForStatement by user or None otherwise.
+ * @param session Spark session that SQL script is executed within.
+ */
+class ForStatementExec(
+    query: SingleStatementExec,
+    variableName: Option[String],
+    body: CompoundBodyExec,
+    val label: Option[String],
+    session: SparkSession) extends NonLeafStatementExec {
+
+  private object ForState extends Enumeration {
+    val VariableAssignment, Body, VariableCleanup = Value
+  }
+  private var state = ForState.VariableAssignment
+  private var areVariablesDeclared = false
+
+  // map of all variables created internally by the for statement
+  // (variableName -> variableExpression)
+  private var variablesMap: Map[String, Expression] = Map()
+
+  // compound body used for dropping variables while in ForState.VariableAssignment
+  private var dropVariablesExec: CompoundBodyExec = null
+
+  private var queryResult: util.Iterator[Row] = _
+  private var isResultCacheValid = false
+  private def cachedQueryResult(): util.Iterator[Row] = {
+    if (!isResultCacheValid) {
+      queryResult = query.buildDataFrame(session).toLocalIterator()
+      query.isExecuted = true
+      isResultCacheValid = true
+    }
+    queryResult
+  }
```

Comment:
awesome, the approach looks good to me now!
it seems like we can simplify this further though, i.e. we don't need a function and helper vars now?

would something simple like this work?
```suggestion
  private lazy val queryResultIter: util.Iterator[Row] = {
    val queryResult = query.buildDataFrame(session).toLocalIterator()
    query.isExecuted = true
    queryResult
  }
```

Reply:
@davidm-db So i tried something like this already, but the problem is that lazy val can't be reset in `reset()`. With lazy val the query will only be executed once, and there is no way to make it execute again in the case the for statement needs to be executed again, for example if it's nested in another loop.

Reply:
ooh, forgot completely about the `reset`... makes sense, thx!